Expose NCCL communicator from ProcessGroupNCCL via an unsafe API

GD06 opened the pull request:
Differential Revision: D73892691




cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k

wconstab commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -2056,6 +2056,11 @@ communication mechanism.
               "_backend_id",
               &::c10d::ProcessGroup::getBackendID,
               py::arg("backend_type"))
+          .def(
```
i think we probably don't want to put this API at the ProcessGroup layer.  Instead we can directly add it at the ProcessGroupNCCL backend.  This avoids the need to have a default impl for gloo and other backends, and the confusion of having a 'general' API for a nccl-specific thing.  

here's an example of adding a nccl-specific API
https://github.com/pytorch/pytorch/blob/main/torch/csrc/distributed/c10d/init.cpp#L3094

and you can see how its used by first calling pg.getBackend() and then calling the api on that
https://github.com/pytorch/pytorch/blob/main/torch/distributed/distributed_c10d.py#L2229

GD06 replied:
Thanks for the suggestion! I reverted changes in the backend, and make this a ProcessGroupNCCL only API.

ngimel commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -2056,6 +2056,11 @@ communication mechanism.
               &::c10d::ProcessGroup::getBackendID,
               py::arg("backend_type"))
+          .def(
+              "_comm_ptr", &::c10d::ProcessGroup::getCommPtr,
```
can you also beef up the warning to not modify or destroy comm?

GD06 replied:
Thx for the suggestion! Updated the warning message.

GD06 committed:
Expose NCCL communicator from ProcessGroupNCCL via an unsafe API (#152496)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/152496

Test Plan: Manual script to test

Reviewed By: ngimel

Differential Revision: D73892691

GD06 commented:
@pytorchbot merge

GD06 commented:
@pytorchbot -f merge

GD06 commented:
@pytorchmergebot  merge -f "CI failure is not caused by this PR."
